docs/src/references/mqtt-api.md

```diff
@@ -0,0 +1,843 @@
+|config_update|`te/<identifier>/cmd/config_update/<cmd_id>`|
+|firmware_update|`te/<identifier>/cmd/firmware_update/<cmd_id>`|
+|restart|`te/<identifier>/cmd/restart/<cmd_id>`|
+
+The command would be interpreted differently based on the target entity.
+For example, the `restart` could mean either a device restart or a service restart based on the target entity.
+
+### Examples: With default device/service topic semantics
+
+#### Command to main device
+
+Command to fetch the software list from the main device:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main///cmd/software_list/123' '{
+  "status": "init"
+}'
+```
+
+The `status` value will transition from `init` to the final `successful` or `failed` via many intermediate states
+such as `validating`, `downloading`, `executing` etc.
+
+The `status` field can even be skipped, which implies `init` status as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main///cmd/software_list/123' '{}'
+```
+
+#### Command to child device
+
+Command to update the firmware of a child device:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/child01///cmd/firmware_update/123' '{
+  "status": "init",
+  "attempt": 1,
+  "name": "OpenWRT",
+  "version": "22.03",
+  "url":"http://127.0.0.1:8000/tedge/file-transfer/tedge-child/firmware_update/93d50a297a8c235",
+  "sha256":"c036cbb7553a909f8b8877d4461924307f27ecb66cff928eeeafd569c3887e29"
+}'
+```
+
+#### Command to a service
+
+Command to update the configuration of a service:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main/service/collectd/cmd/config_update/123' '{
+  "status": "init",
+  "type": "collectd"
+  "path": "/etc/collectd/collectd.conf",
+  "url": "http://127.0.0.1:8000/tedge/file-transfer/collectd/config_update/collectd",
+}'
+```
+
+### Commands metadata
+
+For commands as well, additional command specific metadata can be registered as retained messages on the `<cmd_type>` topic.
+
+For example, the supported configuration list of the main device can be declared as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main///cmd/config_snapshot '{
+  "description": "Upload a configuration from the device",
+  "supportedTypes": ["mosquitto", "tedge", "collectd"]
+}'
+```
+
+## Health check
+
+Services can publish their health status as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main/service/tedge-agent/health/status '{
+  "pid": 1234,
+  "status": "up",
+  "time": 1674739912
+}'
+```
+
+Services are responsible for updating their own health status by publishing to the above topic on any status changes. However, other clients can request the service to update its status by sending a health check command as shown below:
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main/service/tedge-agent/cmd/health/check '{}'
+```
+
+Services are also expected to react to device-wide health check commands as well (where the `<service_id>` is left blank):
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main/service//cmd/health/check '{}'
```

Comment:
```suggestion
tedge mqtt pub -r te/device/main///cmd/health/check '{}'
```
But, I get a feeling that you did that intentionally. Just curious why.

Reply:
Yes this was on purpose. Since we're using the `device/+/service/+` topic schema, being more precise with the wildcard subscription makes more sense to me as blindly listening to anything under the `device/+/+/+` topic might result in some unexpected behaviour. And since the health check is only being used (at the moment) for services, then I thought it best to limit it to services.

Though this logic is no longer applicable when using a custom schema as you need to check for services in the registry and only subscribe to them, or do client side filtering (using its own entity store built from the registration messages).

Reply:
But if @didier-wenzek also prefers (`te/device/main///cmd/health/check` over the `te/device/main/service//cmd/health/check`) then I'm willing to revert the change in topic.

Reply:
> Though this logic is no longer applicable when using a custom schema as you need to check for services in the registry and only subscribe to them, or do client side filtering (using its own entity store built from the registration messages).

This is why I would really avoid to use `te/device/main/`__service__`//cmd/health/check`.  

As the logic is to check *all the services running on a device*, the identifier must be the identifier of a device.
And the software component that ensures these checks must know:
- the device topic identifier (using a config parameter or the default `te/device/main//`.
- the service running to that device (listening to registration messages or requesting the registry).

Reply:
Ok, fair points. I reverted back to `te/device/main///cmd/health/check`.